### PR TITLE
Reformat Docker file for IDR 0.4.5

### PR DESCRIPTION
**Pull request by bramalingam**
Updated the IDR Jupiter docker file to do the following,

1) Change base image (instead of scipy base notebook, start with minimal image)
2) Remove Omego based installation of the whole OMERO server.
3) Use a conda python-omero install 
4) Install Cell Profiler as a python package and r-OMERO dependencies in the same python2 environment.

@joshmoore @jburel @sbesson @manics

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+RUN /opt/conda/envs/python2/bin/pip install \
         graphviz \
         gseapy \
         py2cytoscape \
         pydot \
         tqdm \
-        idr-py==0.1
-
-# Display resource usage in notebooks https://github.com/yuvipanda/nbresuse
-RUN pip install https://github.com/IDR/nbresuse/archive/0.1.0-idr.zip && \
```
Comment (manics) [inline Dockerfile]: Why did you remove `nbresuse`?

File: Dockerfile [hunk 2]
```diff
@@ -1,23 +1,20 @@
-FROM jupyter/scipy-notebook@sha256:4825556416ec7dcf4df73585a71595d29a274ef7d6d7c97267606661f3466952
+FROM jupyter/base-notebook
```
Comment (manics) [inline Dockerfile]: Before this is merged can you use a fixed tag from https://hub.docker.com/r/jupyter/base-notebook/tags/

File: Dockerfile [hunk 3]
```diff
@@ -1,23 +1,20 @@
...
 MAINTAINER ome-devel@lists.openmicroscopy.org.uk
 
+# switch to root suer
```
Comment (manics) [inline Dockerfile]: Typo: `suer` (alternatively delete the unnecessary comments)

File: Dockerfile [hunk 4]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
 # https://github.com/damianavila/RISE
 RUN conda install --quiet --yes -c damianavila82 rise
 
+# install zeroc-ice and python-omero
 RUN conda install --name python2 --quiet --yes -c bioconda zeroc-ice && \
-    conda install --name python2 --quiet --yes -c pdrops pygraphviz && \
-    /opt/conda/envs/python2/bin/pip install \
+    conda install --name python2 --quiet --yes -c bioconda python-omero=5.3.3
```
Comment (manics) [inline Dockerfile]: You can combine these two:
```
conda install --name python2 --quiet --yes -c bioconda \
    zeroc-ice \
    python-omero=5.3.3
```

File: Dockerfile [hunk 5]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+        python-matplotlib  \
+        python-mysqldb     \
+        python-scipy       \
+        python-numpy       \
+        python-wxgtk3.0    \
+        python-zmq
+
+WORKDIR /usr/local/src
+
+# Install CellProfiler
```
Comment (manics) [inline Dockerfile]: Is `/opt/conda/envs/python2/bin/pip install https://github.com/CellProfiler/CellProfiler/archive/v3.0.0.zip` good enough?

File: Dockerfile [hunk 6]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+### --------------
+### See https://github.com/rocker-org/rocker/blob/e9758030e435915d5e6f21aaab0fc35a5a8efaae/r-base/Dockerfile#L41
+### https://github.com/rocker-org/rocker/issues/149
+### --------------
+## Now install R and littler, and create a link for littler in /usr/local/bin
+## Also set a default CRAN repo, and make sure littler knows about it too
+RUN apt-get update \
+        && echo "MIGRATE THIS BLOCK TO THE APT-GET ABOVE" \
+    && apt-get install -y --no-install-recommends \
+        littler \
```
Comment (manics) [inline Dockerfile]: What is `littler` needed for?

File: Dockerfile [hunk 7]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+    && install.r docopt \
+    && rm -rf /tmp/downloaded_packages/ /tmp/*.rds \
+    && rm -rf /var/lib/apt/lists/*
+
+## Install rJava package
+ENV PATH=$PATH:/usr/lib/jvm/java-8-openjdk-amd64/bin
+RUN conda install --quiet --yes -c r r-rjava=0.9_8
+
+# Changed from rOMERO-gateway/Dockerfile
+RUN chown jovyan /usr/local/lib/R/site-library
```
Comment (manics) [inline Dockerfile]: Changing ownership of a system directory seems unusual, is it essential?

File: Dockerfile [hunk 8]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+
+##
+##
+#########################################################################################
+
+USER root
+
+# install romero
+ENV _JAVA_OPTIONS="-Xss2560k -Xmx2g"
+RUN /opt/conda/envs/python2/bin/conda install --quiet --yes -c anaconda gfortran_linux-64
```
Comment (manics) [inline Dockerfile]: What requires `gfortran_linux-64`? Is `anaconda` needed?

File: Dockerfile [hunk 9]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+RUN chown jovyan /usr/local/lib/R/site-library
+
+##
+##
+#########################################################################################
+
+USER root
+
+# install romero
+ENV _JAVA_OPTIONS="-Xss2560k -Xmx2g"
```
Comment (manics) [inline Dockerfile]: Is it appropriate to hard-code memory settings here?

File: Dockerfile [hunk 10]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+    && ln -s /usr/share/doc/littler/examples/install2.r /usr/local/bin/install2.r \
+    && ln -s /usr/share/doc/littler/examples/installGithub.r /usr/local/bin/installGithub.r \
+    && ln -s /usr/share/doc/littler/examples/testInstalled.r /usr/local/bin/testInstalled.r \
+    && install.r docopt \
+    && rm -rf /tmp/downloaded_packages/ /tmp/*.rds \
+    && rm -rf /var/lib/apt/lists/*
+
+## Install rJava package
+ENV PATH=$PATH:/usr/lib/jvm/java-8-openjdk-amd64/bin
+RUN conda install --quiet --yes -c r r-rjava=0.9_8
```
Comment (manics) [inline Dockerfile]: `0.9_9` is available https://anaconda.org/r/r-rjava

File: Dockerfile [hunk 11]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+    && apt-get clean 
+
+ENV JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
+RUN rm -rf /usr/lib/jvm/java
+RUN ln -s  /usr/lib/jvm/java-8-openjdk-amd64 /usr/lib/jvm/java
+
+RUN /opt/conda/envs/python2/bin/conda install --quiet --yes -c r r
+RUN R CMD javareconf
+    
+## make sure Java can be found in rApache and other daemons not looking in R ldpaths
```
Comment (manics) [inline Dockerfile]: Is this needed in this image?

File: Dockerfile [hunk 12]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+## Necessary for running maven
+## copied from https://hub.docker.com/r/cardcorp/r-java/~/dockerfile/
+##
+## gnupg is needed to add new key
+RUN apt-get update && apt-get install -y gnupg2
+
+## Install Java 
+RUN apt-get update \
+    && apt-get install -t jessie-backports -y openjdk-8-jdk \
+    && echo "MAVEN IS NOT IN THE UPSTREAM LIST (JOSH)" \
```
Comment (manics) [inline Dockerfile]: Which version was this written for? `apt-get install openjdk-8-jdk maven` seems to work

File: Dockerfile [hunk 13]
```diff
@@ -39,31 +36,158 @@ RUN conda install --name python2 --quiet --yes \
...
+    && apt-get install -y maven \ 
+    && rm -rf /var/lib/apt/lists/* \
+    && apt-get clean 
+
+ENV JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
+RUN rm -rf /usr/lib/jvm/java
```
Comment (manics) [inline Dockerfile]: `rm -rf ` isn't needed since it doesn't exist

**Review by manics — CHANGES_REQUESTED** [review]
In addition to the in-line comments:
- Can you reduce the number of commands. Each command leads to another layer in the image, which increases the download time
- There's a lot of switching back and for the between different users
- The setup of R seems very complex and difficult to maintain in future. Given that OMERO.py is in bioconda is there any possibility of doing the same for rOMERO?

File: Dockerfile (hunk 1 above)
Comment (bramalingam) [reply Dockerfile]: I had installation issues at this step, while I was working with the Non-IDR version of this docker file for the Cambridge training (because of the base Image I had chosen, happy to add it back and check what the issues were)

File: Dockerfile (hunk 8 above)
Comment (bramalingam) [reply Dockerfile]: Yes this compiler was required for compiling the dependencies for r-java (once the base image was changed)

File: Dockerfile (hunk 1 above)
Comment (manics) [reply Dockerfile]: For purposes of review might it be easier to split this into 3+ PRs?
- minimal image with omero bioconda
- CellProfiler
- R

Comment (bramalingam) [reply Dockerfile]: Happy to do that, if that makes more sense.

File: Dockerfile (hunk 5 above)
Comment (bramalingam) [reply Dockerfile]: I used the docker file snippet that they had shared, https://github.com/CellProfiler/docker. Not sure if the pip installable module takes care of all the dependencies.

File: Dockerfile (hunk 12 above)
Comment (bramalingam) [reply Dockerfile]: @joshmoore : I did not change too many of these steps, when I tried to recover your branch. Happy to discuss and reduce the complexity if need be.

File: Dockerfile (hunk 11 above)
Comment (bramalingam) [reply Dockerfile]: ref : https://cran.r-project.org/doc/manuals/r-release/R-admin.html

"Binary distributions of R are built against a specific version (e.g. 1.8.0_121) of Oracle Java so sudo R CMD javareconf may be needed before using Java-using packages."

File: Dockerfile (hunk 7 above)
Comment (bramalingam) [reply Dockerfile]: I guess this helps in installing R packages from the notebooks? @joshmoore : please correct me if am wrong here.

**Comment by dominikl** [comment]
Once (if) the `romero-gateway` package gets into the CRAN repository, installation will be simpler, a `install.packages('romero-gateway')` call within R should be enough then. Or maybe even via `conda`? Don't know if they somehow automatically include CRAN packages, like they do with `rjava` apparentely.

**Comment by manics** [comment]
What's the timescale for getting it into CRAN? If it's going to be ages we could try setting up our own repo http://kbroman.org/blog/2017/05/23/minicran/ ?

**Comment by dominikl** [comment]
The `minicran` is pretty cool, thanks for the hint @manics . Could we put a minicran repository somewhere under `downloads.openmicroscopy.org` ? Sorry, getting quite off-topic now...

**Review by sbesson — CHANGES_REQUESTED** [review]
Discussed with @bramalingam earlier today. In the scope of IDR 0.4.5, as we have n notebook consuming CellProfiler yet, I would remove this component from the Docker image and limit this PT to to the additions & changes required for fixing the existing notebooks.

@bramalingam mentioned he will split the CellProfiler part and reopen it separately  with a notebook example  but this is outside the scope of the upcoming milestone.

**Comment by jburel** [comment]
CP notebook has only been evaluated during cam2017.
and will need to test it on a larger scale. I agree that it is better not to include it

File: Dockerfile (hunk 10 above)
Comment (bramalingam) [reply Dockerfile]: Version conflicts happening with other r dependencies, when I tried installing 0.9_9!  Will update this with a separate PR, after this is merged.

**Review by sbesson — APPROVED** [review]

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [DOCKER-merge#279](https://ci.openmicroscopy.org/job/DOCKER-merge/279/). See the [console output](https://ci.openmicroscopy.org/job/DOCKER-merge/279/consoleText) for more details.
Possible conflicts:
  - PR #20 manics 'Upgrade to work with jupyter 0.8'
    - Dockerfile

~~--conflicts~~ Conflict resolved in build [DOCKER-merge#282](https://ci.openmicroscopy.org/job/DOCKER-merge/282/). See the [console output](https://ci.openmicroscopy.org/job/DOCKER-merge/282/consoleText) for more details.

**Review by sbesson — CHANGES_REQUESTED** [review]
Discussed today at the IDR meeting. Given some of the implications and the remaining questions to be answered in terms of analytics, the proposal is to split the analytics work as follows:

- for IDR 0.4.5, only fix broken notebooks exclusing the new rOMERO notebook/addition
- for IDR 0.4.6, add the rOMERO notebook and work on the scaling of the analysis using load-balancing + dedicated ports

For this PR, could we move the R-OMERO dependencies section out of this PR for the immediate release and open it separately for the following milestone? This should also address most of @manics' remaining comments in https://github.com/IDR/jupyter-docker/pull/22#pullrequestreview-84144924 which can be discussed in the dedicated PR.

**Review by sbesson — APPROVED** [review]

**Comment by sbesson** [comment]
Retested a few of the notebooks with broken dependencies with the latest version of this Docker image. Everything works as expected. @manics any additional comments?
